includes/class-amp-theme-support.php

```diff
@@ -14,6 +14,84 @@ class AMP_Theme_Support {
 
 	const COMPONENT_SCRIPTS_PLACEHOLDER = '<!--AMP_COMPONENT_SCRIPTS_PLACEHOLDER-->';
 
+	/**
+	 * Template types.
+	 *
+	 * @var array
+	 */
+	protected static $template_types = array(
+		'paged', // Deprecated.
+		'index',
+		'404',
+		'archive',
+		'author',
+		'category',
+		'tag',
+		'taxonomy',
+		'date',
+		'home',
+		'front_page',
+		'page',
+		'search',
+		'single',
+		'embed',
+		'singular',
+		'attachment',
+	);
+
+	/**
+	 * Initialize.
+	 */
+	public static function init() {
+		require_once AMP__DIR__ . '/includes/amp-post-template-actions.php';
+		if ( amp_is_canonical() ) {
+			$is_amp_endpoint = ( false !== get_query_var( AMP_QUERY_VAR, false ) ); // Because is_amp_endpoint() now returns true if amp_is_canonical().
+
+			// Permanently redirect to canonical URL if the AMP URL was loaded, since canonical is now AMP.
+			if ( $is_amp_endpoint ) {
+				wp_safe_redirect( self::get_current_canonical_url(), 301 );
+				exit;
+			}
+		} else {
+			self::register_paired_hooks();
+		}
+		self::register_hooks();
+	}
+
+	/**
+	 * Determines whether paired mode is available.
+	 *
+	 * When 'amp' theme support has not been added or canonical mode is enabled, then this returns false.
+	 * Returns true when there is a template_path defined in theme support, and if a defined active_callback
+	 * returns true.
+	 *
+	 * @return bool Whether available.
+	 */
+	public static function is_paired_available() {
+		$support = get_theme_support( 'amp' );
+		if ( empty( $support ) || amp_is_canonical() ) {
+			return false;
+		}
+
+		$args = array_shift( $support );
+
+		// @todo We might want to rename active_callback to available_callback..
+		if ( isset( $args['active_callback'] ) && is_callable( $args['active_callback'] ) ) {
```

Comment:
Should we rename `active_callback` to `available_callback`? It would seem a little strange for an _active_ callback to return `true` if on the frontend in paired mode, since `is_amp_endpoint()` would be false.

Reply:
I agree the name isn't the most accurate. I've updated it to `available_callback` in b10c0fe, but `display_callback` might even be a more clear. Thoughts?

Reply:
I think I like `available` more. It sounds more like something that would return a boolean. There is a `display_callback` in core, but it is used in widgets to actually render something out. Otherwise, in core, there is `active_callback` and `video-active-callback`.

Core usage: https://gist.github.com/westonruter/49552860d1e57a20547bcebd090ef269

I'm torn between _available_ and _active_.